### PR TITLE
feat: Clipper

### PR DESCRIPTION
**Pull request by dekz**
## Description

```
SELL ETH->USDC 0.16 -> 435.99 ($437.51) after 1.4s @ dev
	✔ PASS 0.16 -> 435.99 ($397.57) ($38.42) [$0.00]
	Clipper: 100%
	gas: 347234 est 345000
SELL DAI->USDC 629008.35 -> 556747.10 ($630053.76) after 2.3s @ dev
	✔ PASS 629008.35 -> 556747.10 ($556711.54) ($35.56) [$0.00]
	Clipper: 100%
	gas: 321433 est 320000
SELL USDC->ETH 402615.72 -> 141.56 ($402615.72) after 4.0s @ dev
	✔ PASS 402615.72 -> 141.56 ($381958.87) ($33.94) [$0.00]
	Clipper: 100%
	gas: 306794 est 345000
```

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: packages/asset-swapper/src/utils/market_operation_utils/sampler_operations.ts [hunk 1]
```diff
@@ -1718,6 +1741,26 @@ export class SamplerOperations {
...
+                        const adjustedTakerToken =
+                            takerToken === NATIVE_FEE_TOKEN_BY_CHAIN_ID[this.chainId] ? NULL_ADDRESS : takerToken;
+                        // Supports the PLP interface
+                        return this.getLiquidityProviderBuyQuotes(
+                            poolAddress,
+                            adjustedMakerToken,
+                            adjustedTakerToken,
+                            makerFillAmounts,
+                            // tslint:disable-next-line: custom-no-magic-numbers
+                            0, // Not used for Clipper
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/utils/market_operation_utils/sampler_operations.ts:1761]: This means their pairs are all roughly equal in gas costs, right?

File: packages/asset-swapper/src/utils/market_operation_utils/orders.ts [hunk 2]
```diff
@@ -475,6 +481,10 @@ export const BRIDGE_ENCODERS: {
     ]),
     [ERC20BridgeSource.KyberDmm]: AbiEncoder.create('(address,address[],address[])'),
     [ERC20BridgeSource.Lido]: AbiEncoder.create('(address)'),
+    [ERC20BridgeSource.Clipper]: AbiEncoder.create([
+        { name: 'provider', type: 'address' },
+        { name: 'data', type: 'bytes' },
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/utils/market_operation_utils/orders.ts:486]: If this is always empty why encode it at all?

File: contracts/zero-ex/contracts/src/transformers/bridges/mixins/MixinClipper.sol [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        internal
+        returns (uint256 boughtAmount)
+    {
+        // Clipper requires ETH and doesn't support WETH
+        WETH.withdraw(sellAmount);
+        boughtAmount = clipper.sellEthForToken{ value: sellAmount }(
+            buyToken,
+            address(this),
+            1,
+            auxiliaryData
```
Comment (dorothy-zbornak) [inline contracts/zero-ex/contracts/src/transformers/bridges/mixins/MixinClipper.sol:98]:
```suggestion
            ""
```
etc

File: contracts/zero-ex/contracts/src/transformers/bridges/mixins/MixinClipper.sol [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        return boughtAmount;
+    }
+
+    function _executeSellEthForToken(
+        ILiquidityProvider clipper,
+        IERC20TokenV06 buyToken,
+        uint256 sellAmount,
+        bytes memory auxiliaryData
+    )
+        internal
```
Comment (dorothy-zbornak) [inline contracts/zero-ex/contracts/src/transformers/bridges/mixins/MixinClipper.sol:89]:
```suggestion
        private
```
etc

**Review by dorothy-zbornak — APPROVED** [review]
small nits but looks good!

File: packages/asset-swapper/src/utils/market_operation_utils/orders.ts (hunk 2 above)
Comment (dekz) [reply packages/asset-swapper/src/utils/market_operation_utils/orders.ts:486]: they want us to send some affiliate data in here, working it out now

File: packages/asset-swapper/src/utils/market_operation_utils/sampler_operations.ts (hunk 1 above)
Comment (dekz) [reply packages/asset-swapper/src/utils/market_operation_utils/sampler_operations.ts:1761]: 1 single pool but we use the gasSchedule for its cost

**Pull request merged**